polyfill/intersection-observer.js

```diff
@@ -0,0 +1,648 @@
+ */
+function computeRectIntersection(rect1, rect2) {
+  var top = Math.max(rect1.top, rect2.top);
+  var bottom = Math.min(rect1.bottom, rect2.bottom);
+  var left = Math.max(rect1.left, rect2.left);
+  var right = Math.min(rect1.right, rect2.right);
+  var intersectionRect = {
+    top: top,
+    bottom: bottom,
+    left: left,
+    right: right,
+    width: left > right ? 0 : right - left,
+    height: top > bottom ? 0 : bottom - top
+  };
+  return intersectionRect;
+}
+
+
+/**
+ * Accepts an intersectionRect and returns whether or not it represents
+ * an actual intersection. This is needed because intersectionRatio alone
+ * doesn't tell the whole story, and there are two cases which it doesn't
+ * account for, both of which should be considered "intersections".
+ * <ul>
+ * <li> A target has an area of zero but is fully in-bounds
+ * <li> A target is on the boarder
+ * </ul>
+ * @param {Object=} opt_rect The intersection rect to check. If no rect is
+ *     provided the result is always false.
+ * @return {boolean} True if an intersection exists, false otherwise.
+ */
+function hasIntersection(opt_rect) {
+  return !!opt_rect &&
+      (opt_rect.left <= opt_rect.right && opt_rect.top <= opt_rect.bottom);
+}
+
+
+/**
+ * Shims the native getBoundingClientRect for compatibility with older IE.
+ * @param {Element} el The element whose bounding rect to get.
+ * @return {Object} The (possibly shimmed) rect of the element.
+ */
+function getBoundingClientRect(el) {
+  var rect = el.getBoundingClientRect();
+  if (!rect) return;
+
+  // Older IE
+  if (!rect.width || !rect.height) {
+    rect = {
+      top: rect.top,
+      right: rect.right,
+      bottom: rect.bottom,
+      left: rect.left,
+      width: rect.right - rect.left,
+      height: rect.bottom - rect.top
+    }
+  }
+  return rect;
+}
+
+
+/**
+ * Returns an empty rect object. An empty rect is returned when an element
+ * is not in the DOM.
+ * @return {Object} The empty rect.
+ */
+function getEmptyRect() {
+  return {
+    top: 0,
+    bottom: 0,
+    left: 0,
+    right: 0,
+    width: 0,
+    height: 0
+  };
+}
+
+
+/**
+ * Determines if a root elements contains a target element as a descendant.
+ * @param {Element} root The root element.
+ * @param {Element} target The target to check.
+ * @return {boolean} True if the target is a descendant of root.
+ */
+function contains(root, target) {
+  var parent = target.parentNode;
+  while (parent) {
+    if (root == parent) return true;
+    parent = parent.parentNode;
+  }
+}
+
+
+/**
+ * Determins if a value is a JavaScript array.
+ * @param {*} value Any JavaScript value.
+ * @return {boolean} True if the passed value is an array.
+ */
+function isArray(value) {
+  return Object.prototype.toString.call(value) == '[object Array]';
+}
+
+
+// Exposes the constructors globally.
+window.IntersectionObserver = IntersectionObserver;
+window.IntersectionObserverEntry = IntersectionObserverEntry;
+
+}(window, document));
```

Comment:
Passing `this` here is not reliable in module loaders like browserify or webpack. In both cases `this` is an empty object in the global scope.

In what cases will `window` not be defined? Specifically, in what cases will `window` not be defined, but `document` (as well other browser-specific globals the script references) will?

I noticed in the polyfill service you use this expression:

```js
'object' === typeof window && window || 'object' === typeof self && self || 'object' === typeof global && global || {}
```

That would be fine with me, but simply using `this` does not work for many cases I need.